docs/proposals/performance-related-monitoring.md

```diff
@@ -0,0 +1,126 @@
+total and average size of them. Each binary should be responsible for exposing its metrics. An unaccounted but way too big number of days, if not weeks, of engineering time was
+lost because of this issue.
+
+### Adding per-pod probe-time, which increased the number of PodStatus updates, causing major slowdown
+
+In September 2015 we tried to add per-pod probe times to the PodStatus. It caused (https://github.com/kubernetes/kubernetes/issues/14273) a massive increase in both number and
+total volume of object (PodStatus) changes. It drastically increased the load on API server which wasn’t able to handle new number of requests quickly enough, violating our
+response time SLO. We had to revert this change.
+
+### Late Ready->Running PodPhase transition caused test failures as it seemed like slowdown
+
+In late September we encountered a strange problem (https://github.com/kubernetes/kubernetes/issues/14554): we observed an increased observed latencies in small clusters (few
+Nodes). It turned out that it’s caused by an added latency between PodRunning and PodReady phases. This was not a real regression, but our tests thought it were, which shows
+how careful we need to be.
+
+### Huge number of handshakes slows down API server
+
+It was a long standing issue for performance and is/was an important bottleneck for scalability. The bug directly causing this problem was incorrect (from the golangs
+standpoint) handling of TCP connections. Secondary issue was that elliptic curve encryption (only one available in go 1.4) is unbelievably slow.
+
+## Proposed metrics/statistics to gather/compute to avoid problems
+
+### Logging monitoring
+
+Log spam is a serious problem and we need to keep it under control. Simplest way to check for regressions, suggested by @bredanburns, is to compute the rate in which log files
+grow in e2e tests.
+
+Basic ideas:
+- log generation rate (B/s)
+
+### REST call monitoring
+
+We do measure REST call duration in the Density test, but we need an API server monitoring as well, to avoid false failures caused e.g. by the network traffic. We already have
+some metrics in place (https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/metrics/metrics.go), but we need to revisit the list and add some more.
+
+Basic ideas:
+- number of calls per verb, client, resource type
+- latency distribution per verb, client, resource type
+- number of calls that was rejected per client, resource type and reason (invalid version number, already at maximum number of requests in flight)
+
+### Rate limit monitoring
+
+Reverse of REST call monitoring done in the API server. We need to know when a given component increases a pressure it puts on the API server. As a proxy for number of
+requests sent we can track how saturated are rate limiters. This has additional advantage of giving us data needed to fine-tune rate limiter constants.
```

Comment:
https://github.com/kubernetes/kubernetes/issues/14955 mentions possibly replacing client limits with server-side throttling.  I know the API server currently has a `max-requests-inflight` parameter, but I don't know whether we have a metric to cover how close we are to hitting that limit.  Seems like our best proxy for that is a spike in occurrence of 429 responses?